test-config.md

```diff
@@ -0,0 +1,56 @@
+
+1. When the actor executes the `createGroup` action, a client with a session is created.
+2. When the actor executes the `createKeyPackage` action, a client without a session is created. When later the actor executes the joinGroup action, a session for the client is created.
+
+Note: Creating a new client / session silently overwrites the first one.
+
+
+### Actions
+
+Each action must specify two values: `action` indicating the action type (e.g. `removeProposal`) and `actor` identifying the actor performing the action. Some actions require additional values, e.g., the leaf index of a removed member or a pointer to a packet generated as a result of a previous action. A pointers to the outcome of a previous action is always its index in the sequence of all actions in the given script.
+
+The outputs of all actions are logged in the transcript. They can be either packets packets sent to the delivery service (logged as hex-encoded bytes) or values outputted to the user (stored as strings). The transcript is printed to STDOUT.
+
+
+The following actions are supported.
+
+
+| `action`             | Values specified by the action in addition to `actor` and `action` | Output values stored in the transcript | Description                                       |
+|:---------------------|:-------------------------------------------|:------------------------------|:--------------------------------------------------|
+| `createGroup`        | -                                          | -                             | A new client for the `actor` and a new group with themselves in it are created.      |
+| `createKeyPackage`   | -                                          | `keyPackage`: packet                   | A new client for the `actor` and a new key package for them are created.      |
+| `addProposal`        | `keyPackage`: int                          | `proposal`: packet               | The `actor`’s current client sends a proposal that adds a member with the key package created during the `createKeyPackage` action with index `keyPackage`. |
+| `removeProposal`     | `removed`: string                          | `proposal`: packet               | The `actor`’s current client sends a proposal that removes the actor indicated in `removed`. |
+| `updateProposal`     | -                                          | `proposal`: packet               | The `actor`’s current client sends an update. |
+| `commit`             | `byReference`: []int                       | `commit`: packet, `welcome`: packet | The `actor`’s current client first processes the proposals created during actions identified by the indices in the list `byReference` and then they send a commit. The `byReference` list MUST NOT include proposals from `actor` (those should be committed automatically). |
```

Comment:
The reason is that I wasn't sure how to interpret this input. Proposals by value are new proposals from the committer, so how should the test client interpret a proposal packet as a new thing to propose? Extracting this information is not something a client will normally do, so it may require hacks in some implementations.

However, I don't see how to improve the test runner without making it too complicated, so I guess this will have to do.

Reply:
Yeah, I was thinking that you would generate full, signed proposals, then the committer would just rip the `Proposal` out of the middle and re-originate it.  It's kind of a hack.

You could make it a little more elegant in the following way:
* Add a `bool signed` parameter to the RPCs to get proposals
* When `signed`, emit an MLSPlaintext / MLSCiphertext
* When `!signed`, emit a Proposal
* Gather Proposals from the committer for use by value

This would work OK against the current MLSpp API, not sure how it looks wrt the others.